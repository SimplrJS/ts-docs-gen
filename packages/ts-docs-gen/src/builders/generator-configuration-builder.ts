import * as ts from "typescript";
import * as path from "path";
import * as fs from "fs-extra";
import { Extractor, GetCompilerOptions, Contracts } from "ts-extractor";
import { ApiHelpers } from "ts-extractor/dist/internal";

import { GeneratorConfiguration, WorkingGeneratorConfiguration } from "../contracts/generator-configuration";
import { Plugin } from "../contracts/plugin";

import { PluginRegistry } from "../registries/plugin-registry";
import { DefaultPlugins } from "../default-plugins";
import { GeneratorHelpers } from "../generator-helpers";
import { LoggerHelpers } from "../utils/logger";

export class GeneratorConfigurationBuilder {
    constructor(private projectDirectory: string) {
        this.configuration.projectDirectory = projectDirectory;
    }

    private configuration: Partial<WorkingGeneratorConfiguration> = {
        excludePrivateApi: true
    };
    private compilerOptions: Partial<ts.CompilerOptions> | undefined;
    private tsConfigLocation: string | undefined;

    private resolveProjectDirectory(): string {
        return this.configuration.projectDirectory || this.projectDirectory;
    }

    private extractorFilterApiItem: Contracts.FilterApiItemsHandler = (apiItem): boolean => {
        // Exclude private api.
        if (this.configuration.excludePrivateApi) {
            // Check Access Modifier.
            const accessModifier = ApiHelpers.ResolveAccessModifierFromModifiers(apiItem.Declaration.modifiers);
            if (accessModifier === Contracts.AccessModifier.Private) {
                return false;
            }

            // Look for JSDocTag "@private"
            const metadata = apiItem.GetItemMetadata();
            if (metadata.JSDocTags.findIndex(x => x.name === GeneratorHelpers.JSDocTags.Private) !== -1) {
                return false;
            }
        }
        return true;
    }

    /**
     * @param tsconfigLocation Relative `tsconfig.json` location from `projectDirectory`.
     */
    public SetTsConfigLocation(tsconfigLocation: string): this {
        this.tsConfigLocation = tsconfigLocation;

        return this;
    }

    public OverrideCompilerOptions(compilerOptions: Partial<ts.CompilerOptions>): this {
        this.compilerOptions = {
            ...this.compilerOptions,
            ...compilerOptions
        };

        return this;
    }

    /**
     * Override configuration with a new configuration object.
     *
     * @param configuration Partial configuration object.
     */
    public OverrideConfiguration(configuration: Partial<WorkingGeneratorConfiguration>): this {
        this.configuration = {
            ...this.configuration,
            ...configuration
        };

        return this;
    }

    public SetExternalPackages(packages: string[]): this {
        this.configuration.externalPackage = packages;

        return this;
    }

    public SetOutputDirectory(outputDirectory: string): this {
        this.configuration.outputDirectory = outputDirectory;

        return this;
    }

    public AddPlugins(plugins: Plugin[]): this {
        const currentPlugins = this.configuration.plugins || [];
        this.configuration.plugins = [...plugins, ...currentPlugins];

        return this;
    }

    public async Build(entryFiles: string[]): Promise<GeneratorConfiguration> {
        // Verbosity level.
        if (this.configuration.verbosity != null) {
            LoggerHelpers.SetLogLevel(this.configuration.verbosity);
        }

        // Register all plugins.
        const pluginManager = new PluginRegistry();
        // Register default plugins
        for (const item of DefaultPlugins) {
            pluginManager.Register(item);
        }

        if (this.configuration.plugins != null) {
            // Registering given plugins.
            for (const plugin of this.configuration.plugins) {
                pluginManager.Register(plugin);
            }
        }

        // Resolve tsconfig
        let compilerOptions = this.compilerOptions;
        if (compilerOptions == null) {
            const tsconfigLocation = this.tsConfigLocation || path.join(this.resolveProjectDirectory(), "tsconfig.json");
            compilerOptions = await GetCompilerOptions(tsconfigLocation);
        }

        // Extractor
        const extractor = new Extractor({
            CompilerOptions: compilerOptions,
            ProjectDirectory: this.resolveProjectDirectory(),
            Exclude: this.configuration.exclude,
            OutputPathSeparator: this.configuration.outputPathSeparator,
            ExternalPackages: this.configuration.externalPackage,
            FilterApiItems: this.extractorFilterApiItem,
            Verbosity: this.configuration.verbosity
        });

        const extractedData = extractor.Extract(entryFiles);
        // extractorOutput
        if (this.configuration.extractorOutput) {
            await fs.writeJson(this.configuration.extractorOutput, extractedData, { spaces: 4 });
        }

        // Output directory
        const outputDirectory = this.configuration.outputDirectory || path.join(this.resolveProjectDirectory(), "/docs/");

        return {
            PluginManager: pluginManager,
            ExtractedData: extractedData,
            OutputDirectory: outputDirectory
        };
    }
}
