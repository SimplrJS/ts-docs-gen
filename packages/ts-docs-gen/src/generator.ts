import { Contracts } from "ts-extractor";
import * as path from "path";
import * as fs from "fs-extra";

import { GeneratorConfiguration } from "./contracts/generator-configuration";
import { FileManager } from "./file-manager";
import { RenderItemOutputDto } from "./contracts/render-item-output-dto";
import { ReferenceTuple } from "./contracts/reference-tuple";
import { ApiDefaultPlugin } from "./plugins/api-default-plugin";

import { ExtractorHelpers } from "./extractor-helpers";
import { FileOutputDto } from "./contracts/file-output-dto";
import { PluginData } from "./contracts/plugin-data";

export class Generator {
    constructor(private configuration: GeneratorConfiguration) {
        this.fileManager = new FileManager();
        const { ExtractedData } = this.configuration;

        for (const entryFile of this.configuration.ExtractedData.EntryFiles) {
            const referenceTuples = ExtractorHelpers.GetReferenceTuples(ExtractedData, entryFile.Members);

            for (const reference of referenceTuples) {
                const [referenceId] = reference;

                const renderedItem = this.getRenderedItemByReference(reference);
                this.fileManager.AddItem(renderedItem, referenceId, this.getFilePathFromEntryFile(entryFile));
            }
        }

        this.outputData = this.fileManager.ToFilesOutput();
    }

    private renderedItems: Map<ReferenceTuple, RenderItemOutputDto> = new Map();
    private fileManager: FileManager;
    private outputData: FileOutputDto[];

    private getFilePathFromEntryFile(entryFile: Contracts.ApiSourceFileDto): string {
        return path.basename(entryFile.Name, path.extname(entryFile.Name)) + ".md";
    }

    public get OutputData(): ReadonlyArray<FileOutputDto> {
        return this.outputData;
    }

    public async WriteToFiles(): Promise<void> {
        for (const item of this.outputData) {
            const fullLocation = path.join(this.configuration.OutputDirectory, "api", item.FileLocation);

            try {
                // Ensure output directory
                await fs.ensureDir(path.dirname(fullLocation));
                // Output file
                await fs.writeFile(fullLocation, item.Output.join("\n"));
            } catch (error) {
                console.error(error);
            }
        }
    }

    private getRenderedItemByReference = (reference: ReferenceTuple): RenderItemOutputDto => {
        const [referenceId] = reference;
        const renderedItem = this.renderedItems.get(reference);

        if (renderedItem == null) {
            const { Registry } = this.configuration.ExtractedData;
            const renderedData = this.renderApiItem(reference, Registry[referenceId]);
            this.renderedItems.set(reference, renderedData);

            return renderedData;
        }

        return renderedItem;
    }

    private renderApiItem(
        reference: ReferenceTuple,
        apiItem: Contracts.ApiItemDto,
        parentId?: string
    ): RenderItemOutputDto {
        const plugins = this.configuration.PluginManager.GetPluginsByKind(apiItem.ApiKind);

        const pluginData: PluginData = {
            ExtractedData: this.configuration.ExtractedData,
            Reference: reference,
            ApiItem: apiItem,
            GetItem: this.getRenderedItemByReference,
            ParentId: parentId
        };

        for (const plugin of plugins) {
            if (plugin.CheckApiItem(apiItem)) {
                return plugin.Render(pluginData);
            }
        }

        const defaultPlugin = new ApiDefaultPlugin();
        return defaultPlugin.Render(pluginData);
    }
}
