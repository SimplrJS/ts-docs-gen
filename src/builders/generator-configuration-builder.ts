import * as ts from "typescript";
import * as path from "path";
import { Extractor, GetCompilerOptions } from "ts-extractor";

import { ApiItemPluginBase } from "../abstractions/api-item-plugin-base";
import { GeneratorConfiguration, WorkingGeneratorConfiguration } from "../contracts/generator-configuration";

import { PluginManager } from "../managers/plugin-manager";

// TODO: Add method to read compiler options from tsconfig.
// TODO: Configuration Updater.
export class GeneratorConfigurationBuilder {
    constructor(private projectDirectory: string) { }

    private configuration: Partial<WorkingGeneratorConfiguration> = {};
    private compilerOptions: Partial<ts.CompilerOptions>;

    public OverrideCompilerOptions(compilerOptions: Partial<ts.CompilerOptions>): this {
        this.compilerOptions = {
            ...this.compilerOptions,
            ...compilerOptions
        };

        return this;
    }

    /**
     * Override configuration with new configuration object.
     *
     * @param configuration Partial configuration object.
     */
    public OverrideConfiguration(configuration: Partial<WorkingGeneratorConfiguration>): this {
        this.configuration = {
            ...this.configuration,
            ...configuration
        };

        return this;
    }

    public SetOutputDirectory(outputDirectory: string): this {
        this.configuration.OutputDirectory = outputDirectory;

        return this;
    }

    public AddPlugins(plugins: ApiItemPluginBase[]): this {
        const currentPlugins = this.configuration.Plugins || [];
        this.configuration.Plugins = [...plugins, ...currentPlugins];

        return this;
    }

    public async Build(entryFiles: string[]): Promise<GeneratorConfiguration> {
        // Register all plugins.
        const pluginManager = new PluginManager();
        if (this.configuration.Plugins != null) {
            // TODO: Register default plugins.
            // Registering given plugins.
            for (const plugin of this.configuration.Plugins) {
                pluginManager.Register(plugin);
            }

        }

        // Resolve tsconfig
        let compilerOptions = this.compilerOptions;
        if (compilerOptions == null) {
            compilerOptions = await GetCompilerOptions(path.join(this.projectDirectory, "tsconfig.json"));
        }

        // Extractor
        const extractor = new Extractor({
            CompilerOptions: compilerOptions,
            ProjectDirectory: this.projectDirectory
        });

        const outputDirectory = this.configuration.OutputDirectory || path.join(this.projectDirectory, "/docs/");

        return {
            EntryFiles: entryFiles,
            PluginManager: pluginManager,
            ExtractedData: extractor.Extract(entryFiles),
            OutputDirectory: outputDirectory
        };
    }
}


const output = new GeneratorConfigurationBuilder("./")
    .AddPlugins([])
    .OverrideCompilerOptions({ typeRoots: ["."] })
    .SetOutputDirectory("./docs")
    .Build(["./src/index.ts", "./src/contracts.ts"]);
