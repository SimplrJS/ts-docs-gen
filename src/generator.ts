import { GeneratorConfiguration } from "./contracts/generator-configuration";
import { RenderItemOutputDto } from "./contracts/render-item-output-dto";
import { Contracts } from "ts-extractor";

// TODO: Move to contracts.
export type RenderedItems = Map<string, RenderItemOutputDto>;

export interface RenderedDto {
    RenderedItems: RenderedItems;
    EntryFiles: Contracts.ApiSourceFileDto[];
}

export class Generator {
    constructor(private configuration: GeneratorConfiguration) { }

    private renderedItems: Map<string, RenderItemOutputDto> = new Map();
    private renderedData: RenderedDto | undefined;

    private renderApiItem(apiItem: Contracts.ApiItemDto): RenderItemOutputDto {
        const plugins = this.configuration.PluginManager.GetPluginsByKind(apiItem.ApiKind);

        for (const plugin of plugins) {
            if (plugin.CheckApiItem(apiItem)) {
                return plugin.Render(apiItem, this.getRenderedItemById);
            }
        }

        // FIXME: Implement default plugin.
        return undefined as any;
    }

    // TODO: Check for infinity loop.
    private getRenderedItemById = (itemId: string): RenderItemOutputDto => {
        if (!this.renderedItems.has(itemId)) {
            const { Registry } = this.configuration.ExtractedData;
            const renderedData = this.renderApiItem(Registry[itemId]);
            this.renderedItems.set(itemId, renderedData);

            return renderedData;
        }

        return this.renderedItems.get(itemId)!;
    }

    private onRenderData(): RenderedDto {
        const { Registry, EntryFiles } = this.configuration.ExtractedData;

        for (const [itemKey] of Object.entries(Registry)) {
            if (!this.renderedItems.has(itemKey)) {
                this.getRenderedItemById(itemKey);
            }
        }

        return {
            EntryFiles: EntryFiles,
            RenderedItems: this.renderedItems
        };
    }

    public GetRenderedData(): RenderedDto {
        let data: RenderedDto | undefined = this.renderedData;
        if (data == null) {
            data = this.onRenderData();
        }

        return data;
    }

    public PrintToFiles(): void {
        throw new Error("Not yet implemented!");
    }
}
