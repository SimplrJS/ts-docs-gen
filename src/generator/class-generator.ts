import * as json2md from "json2md";

import {
    ApiJson,
    Members,
    MemberClass,
    MemberProperty,
    MemberMethod
} from "../extractor/api-json-contracts";
import { HelpersGenerator } from "./helpers-generator";
import { PropertyGenerator } from "./property-generator";
import { MethodGenerator } from "./method-generator";

export namespace ClassGenerator {
    export function RenderClass(name: string, memberClass: MemberClass): json2md.DataObject[] {
        let md: json2md.DataObject[] = [
            {
                h3: name
            }
        ];

        // Summary
        if (memberClass.summary.length !== 0) {
            md.push({
                p: memberClass.summary.map(x => x.value).join("\n")
            });
        }

        // Remarks
        if (memberClass.remarks.length !== 0) {
            md.push({
                p: memberClass.remarks.map(x => x.value).join("\n")
            });
        }

        // Extends
        if (memberClass.extends !== "") {
            md = md.concat(HelpersGenerator.RenderExtends(memberClass.extends));
        }

        // Implements
        if (memberClass.implements !== "") {
            md = md.concat(HelpersGenerator.RenderImplements(memberClass.implements));
        }

        const properties: { [name: string]: MemberProperty } = {};
        const methods: { [name: string]: MemberMethod } = {};
        for (const memberKey in memberClass.members) {
            if (memberClass.members.hasOwnProperty(memberKey)) {
                const member = memberClass.members[memberKey];
                switch (member.kind) {
                    case "method":
                        methods[memberKey] = member;
                        break;
                    case "property":
                        properties[memberKey] = member;
                        break;
                }
            }
        }

        md = md.concat(renderProperties(properties));
        md = md.concat(renderMethods(methods));

        return md;
    }

    function renderProperties(properties: { [name: string]: MemberProperty }): json2md.DataObject[] {
        let md: json2md.DataObject[] = [
            {
                h4: "Properties"
            }
        ];

        let list: json2md.DataObject[] = [];

        for (const propertyName in properties) {
            if (properties.hasOwnProperty(propertyName)) {
                const memberProperty = properties[propertyName];
                list = list.concat(PropertyGenerator.RenderProperty(propertyName, memberProperty));
            }
        }

        if (list.length === 0) {
            return [];
        }

        md = md.concat(list);

        return md;
    }

    function renderMethods(methods: { [name: string]: MemberMethod }): json2md.DataObject[] {
        let md: json2md.DataObject[] = [
            {
                h4: "Methods"
            }
        ];

        let list: json2md.DataObject[] = [];

        for (const methodName in methods) {
            if (methods.hasOwnProperty(methodName)) {
                const memberMethod = methods[methodName];
                list = list.concat(MethodGenerator.RenderMethod(methodName, memberMethod));
            }
        }

        if (list.length === 0) {
            return [];
        }

        md = md.concat(list);

        return md;
    }
}
