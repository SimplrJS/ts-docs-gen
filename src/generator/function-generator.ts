import * as json2md from "json2md";

import { MemberFunction, Param, ReturnValue } from "../extractor/api-json-contracts";
import { HelpersGenerator } from "./helpers-generator";

export namespace FunctionGenerator {
    export function RenderFunction(name: string, memberFunction: MemberFunction): json2md.DataObject[] {
        let md: json2md.DataObject[] = [
            {
                h3: name
            }
        ];

        // Summary
        if (memberFunction.summary.length !== 0) {
            md.push({
                p: memberFunction.summary.map(x => x.value).join("\n")
            });
        }

        // Remarks
        if (memberFunction.remarks.length !== 0) {
            md.push({
                p: memberFunction.remarks.map(x => x.value).join("\n")
            });
        }

        md.push({
            code: {
                language: "ts",
                content: memberFunction.declarationLine.replace("export ", "")
            }
        });

        // Parameters
        md = md.concat(renderFunctionParameters(memberFunction.parameters));

        // Returns value
        const returnValueDesc = memberFunction.returnValue.description.map(x => x.value).join(".");
        md = md.concat([
            {
                h4: "Returns"
            },
            {
                p: `${HelpersGenerator.InlineCode(memberFunction.returnValue.type)} ${returnValueDesc}`
            }
        ]);

        return md;
    }



    function renderFunctionParameters(parameters: { [key: string]: Param }): json2md.DataObject[] {
        const md: json2md.DataObject[] = [
            {
                h4: "Parameters"
            }
        ];
        const parametersList: string[] = [];

        for (const parameterName in parameters) {
            if (parameters.hasOwnProperty(parameterName)) {
                const parameter = parameters[parameterName];
                const optional = (parameter.isOptional ? "?" : "");
                const line: string = `${parameterName}${optional}: ${HelpersGenerator.InlineCode(parameter.type)}`;

                // FIXME: When there is description on this param.
                // if (parameter.description != null) {
                //     // line += ` - ${parameter.description.map(x => x.}` 
                // }

                parametersList.push(line);
            }
        }

        if (parametersList.length === 0) {
            return [];
        }

        md.push({
            ul: parametersList
        });

        return md;
    }
}
