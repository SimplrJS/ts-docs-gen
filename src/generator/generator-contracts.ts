import * as json2md from "json2md";

export interface Table extends json2md.DefaultConverters.TableInput {
    headers: string[];
    rows: string[][];
}
