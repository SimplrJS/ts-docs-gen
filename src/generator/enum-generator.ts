import * as json2md from "json2md";
import { MemberEnum, MemberEnumValue } from "../extractor/api-json-contracts";
import { HelpersGenerator } from "./helpers-generator";
import { Table } from "./generator-contracts";

export namespace EnumGenerator {
    export function RenderEnum(name: string, memberEnum: MemberEnum): json2md.DataObject[] {
        const md: json2md.DataObject[] = [
            {
                h3: name
            }
        ];

        // Summary
        if (memberEnum.summary.length !== 0) {
            md.push({
                p: memberEnum.summary.map(x => x.value)
            });
        }

        // Remarks
        if (memberEnum.remarks.length !== 0) {
            md.push({
                p: memberEnum.remarks.map(x => x.value)
            });
        }

        // Table of values
        md.push({
            table: renderEnumValues(memberEnum.values)
        });

        return md;
    }

    function renderEnumValues(members: { [member: string]: MemberEnumValue }): Table {
        const table: Table = {
            headers: ["Name", "Value", "Summary"],
            rows: []
        };

        for (const memberKey in members) {
            if (members.hasOwnProperty(memberKey)) {
                const member = members[memberKey];
                const row: string[] = [
                    memberKey,
                    member.value,
                    member.summary.map(x => x.value).join(".")
                ];

                table.rows.push(row);
            }
        }

        // Sort rows by values
        table.rows.sort((x, y) => {
            const xValue = x[1];
            const yValue = y[1];
            if (isNumberCheck(xValue) && isNumberCheck(yValue)) {
                return xValue - yValue;
            }

            return 0;
        });

        return table;
    }

    function isNumberCheck(a: string | number): a is number {
        return isFinite(a as number);
    }
}
