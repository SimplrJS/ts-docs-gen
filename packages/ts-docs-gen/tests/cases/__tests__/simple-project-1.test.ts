import { Generator } from "@src/generator";
import { GeneratorConfigurationBuilder } from "@src/builders/generator-configuration-builder";

test("index", async done => {
    const projectDirectory = "D:/Projects/ts-docs-gen/packages/ts-docs-gen/tests/cases/simple-project-1";
    const entryFiles  = ["./index.ts"];

    try {
        const configuration = await new GeneratorConfigurationBuilder(projectDirectory)
            .Build(entryFiles);

        const generator = new Generator(configuration);

        expect(generator.OutputData).toMatchSnapshot();
        done();
    } catch (error) {
        done.fail(error);
    }
});
