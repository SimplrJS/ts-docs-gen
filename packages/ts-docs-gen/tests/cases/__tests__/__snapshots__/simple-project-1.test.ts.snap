// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple-project-1 1`] = `
Array [
  Object {
    "FileLocation": "index.md",
    "Result": Array [
      "[ClassDeclaration-0]: index/foo.md#class-foo",
      "[ClassDeclaration-1]: index/world.md#class-world",
      "[ClassDeclaration-2]: index/earth.md#class-earth",
      "# index",
      "",
      "## EnumList",
      "",
      "Simple list.",
      "",
      "",
      "\`\`\`typescript",
      "enum EnumList {",
      "     FirstOption = 0,",
      "     SecondOption = 1,",
      "     ThirdOption = 2",
      "}",
      "\`\`\`",
      "",
      "| Name         | Value | Description                    |",
      "| ------------ | ----- | ------------------------------ |",
      "| FirstOption  | 0     | Description for First option.  |",
      "| SecondOption | 1     | Description for Second option. |",
      "| ThirdOption  | 2     | Description for Third option.  |",
      "",
      "## EnumListWithNumberValues",
      "",
      "List with number values with no punctuation at the end of description",
      "",
      "",
      "\`\`\`typescript",
      "enum EnumListWithNumberValues {",
      "     FirstOption = 1,",
      "     SecondOption = 2,",
      "     ThirdOption = 3",
      "}",
      "\`\`\`",
      "",
      "| Name         | Value |",
      "| ------------ | ----- |",
      "| FirstOption  | 1     |",
      "| SecondOption | 2     |",
      "| ThirdOption  | 3     |",
      "",
      "## EnumListWithStringValues",
      "",
      "<span style=\\"color: #d2d255;\\">Warning: Beta!</span>",
      "",
      "<span style=\\"color: red;\\">Deprecated!</span>",
      "",
      "",
      "\`\`\`typescript",
      "enum EnumListWithStringValues {",
      "     FirstOption = \\"first\\",",
      "     SecondOption = \\"second\\",",
      "     ThirdOption = \\"third\\"",
      "}",
      "\`\`\`",
      "",
      "| Name         | Value    |",
      "| ------------ | -------- |",
      "| FirstOption  | \\"first\\"  |",
      "| SecondOption | \\"second\\" |",
      "| ThirdOption  | \\"third\\"  |",
      "",
      "## SampleConst",
      "",
      "\`\`\`typescript",
      "const SampleConst: string;",
      "\`\`\`",
      "",
      "### Type",
      "",
      "string",
      "",
      "## [Foo][ClassDeclaration-0]",
      "",
      "## [World][ClassDeclaration-1]",
      "",
      "## [Earth][ClassDeclaration-2]",
      "",
      "## AnotherFoo(parameter1, parameter2)",
      "",
      "\`\`\`typescript",
      "function AnotherFoo<T extends Object>(parameter1: string, parameter2: Promise<T>): string",
      "\`\`\`",
      "",
      "### Type parameters",
      "",
      "| Name | Constraint type |",
      "| ---- | --------------- |",
      "| T    | Object          |",
      "",
      "### Parameters",
      "",
      "| Name       | Type    |",
      "| ---------- | ------- |",
      "| parameter1 | string  |",
      "| parameter2 | Promise |",
      "",
      "### Return type",
      "",
      "string",
      "",
      "",
      "## GetFoo()",
      "",
      "\`\`\`typescript",
      "function GetFoo(): Promise<void>",
      "\`\`\`",
      "",
      "### Return type",
      "",
      "Promise",
      "",
      "",
      "## FunctionWithOneParameter(parameter)",
      "",
      "\`\`\`typescript",
      "function FunctionWithOneParameter(parameter: string): void",
      "\`\`\`",
      "",
      "### Parameters",
      "",
      "| Name      | Type   |",
      "| --------- | ------ |",
      "| parameter | string |",
      "",
      "### Return type",
      "",
      "void",
      "",
      "",
      "## FunctionWithNoParameters()",
      "",
      "\`\`\`typescript",
      "function FunctionWithNoParameters(): void",
      "\`\`\`",
      "",
      "### Return type",
      "",
      "void",
      "",
      "",
      "## FunctionWithMultipleParameters(parameter1, parameter2)",
      "",
      "\`\`\`typescript",
      "function FunctionWithMultipleParameters(parameter1: string, parameter2: number): void",
      "\`\`\`",
      "",
      "### Parameters",
      "",
      "| Name       | Type   |",
      "| ---------- | ------ |",
      "| parameter1 | string |",
      "| parameter2 | number |",
      "",
      "### Return type",
      "",
      "void",
      "",
      "",
      "## Bar(parameter1, parameter2)",
      "",
      "<span style=\\"color: #d2d255;\\">Warning: Beta!</span>",
      "",
      "<span style=\\"color: red;\\">Deprecated!</span>",
      "",
      "Some general comment about Bar function.",
      "",
      "\`\`\`typescript",
      "function Bar(parameter1: string, parameter2: number): string",
      "\`\`\`",
      "",
      "### Parameters",
      "",
      "| Name       | Type   |",
      "| ---------- | ------ |",
      "| parameter1 | string |",
      "| parameter2 | number |",
      "",
      "### Return type",
      "",
      "string",
      "",
      "",
      "## FunctionWithMultipleTypeParameters(parameter1, parameter2)",
      "",
      "Comment on Function with multiple type parameters.",
      "",
      "\`\`\`typescript",
      "function FunctionWithMultipleTypeParameters<T extends Object, P>(parameter1: T, parameter2: P): string",
      "\`\`\`",
      "",
      "### Type parameters",
      "",
      "| Name | Constraint type |",
      "| ---- | --------------- |",
      "| T    | Object          |",
      "| P    |                 |",
      "",
      "### Parameters",
      "",
      "| Name       | Type |",
      "| ---------- | ---- |",
      "| parameter1 | T    |",
      "| parameter2 | P    |",
      "",
      "### Return type",
      "",
      "string",
      "",
      "",
      "## FunctionWithTypeParameterDefault(parameter1, parameter2)",
      "",
      "Some general comment about Function with type parameter default function.",
      "",
      "\`\`\`typescript",
      "function FunctionWithTypeParameterDefault<T extends Object = {}>(parameter1: string, parameter2: T): string",
      "\`\`\`",
      "",
      "### Type parameters",
      "",
      "| Name | Constraint type | Default type |",
      "| ---- | --------------- | ------------ |",
      "| T    | Object          | \\\\{\\\\}         |",
      "",
      "### Parameters",
      "",
      "| Name       | Type   |",
      "| ---------- | ------ |",
      "| parameter1 | string |",
      "| parameter2 | T      |",
      "",
      "### Return type",
      "",
      "string",
      "",
      "",
      "## FunctionWithTypeParameterConstraint(parameter1, parameter2)",
      "",
      "Some general comment about AnotherBar function.",
      "",
      "\`\`\`typescript",
      "function FunctionWithTypeParameterConstraint<T extends Object>(parameter1: string, parameter2: T): string",
      "\`\`\`",
      "",
      "### Type parameters",
      "",
      "| Name | Constraint type |",
      "| ---- | --------------- |",
      "| T    | Object          |",
      "",
      "### Parameters",
      "",
      "| Name       | Type   |",
      "| ---------- | ------ |",
      "| parameter1 | string |",
      "| parameter2 | T      |",
      "",
      "### Return type",
      "",
      "string",
      "",
      "",
      "## FunctionWithoutReturnType(parameter1, parameter2)",
      "",
      "\`\`\`typescript",
      "function FunctionWithoutReturnType<T extends T[]>(parameter1: string, parameter2: Promise<T>): string",
      "\`\`\`",
      "",
      "### Type parameters",
      "",
      "| Name | Constraint type |",
      "| ---- | --------------- |",
      "| T    | Array           |",
      "",
      "### Parameters",
      "",
      "| Name       | Type    |",
      "| ---------- | ------- |",
      "| parameter1 | string  |",
      "| parameter2 | Promise |",
      "",
      "### Return type",
      "",
      "string",
      "",
      "",
      "## FunctionWithGenericReturnType()",
      "",
      "\`\`\`typescript",
      "function FunctionWithGenericReturnType(): string[]",
      "\`\`\`",
      "",
      "### Return type",
      "",
      "Array",
      "",
      "",
      "## FunctionWithPrimitiveReturnType()",
      "",
      "\`\`\`typescript",
      "function FunctionWithPrimitiveReturnType(): boolean",
      "\`\`\`",
      "",
      "### Return type",
      "",
      "true | false",
      "",
      "",
      "## FunctionWithUnionReturnType()",
      "",
      "\`\`\`typescript",
      "function FunctionWithUnionReturnType(): \\"something\\" | \\"nothing\\"",
      "\`\`\`",
      "",
      "### Return type",
      "",
      "\\"something\\" | \\"nothing\\"",
      "",
      "",
      "## FunctionWithIntersectionReturnType()",
      "",
      "\`\`\`typescript",
      "function FunctionWithIntersectionReturnType(): Earth & World",
      "\`\`\`",
      "",
      "### Return type",
      "",
      "[Earth][ClassDeclaration-2] & [World][ClassDeclaration-1]",
      "",
      "",
      "## MyFunction()",
      "",
      "Function with TypeParameter as TypeLiteral.",
      "",
      "\`\`\`typescript",
      "function MyFunction<T extends { myProperty: string; myPropertyTwo?: number | undefined; } = { myProperty: string; }>(): T",
      "\`\`\`",
      "",
      "### Type parameters",
      "",
      "| Name | Constraint type                                                | Default type              |",
      "| ---- | -------------------------------------------------------------- | ------------------------- |",
      "| T    | \\\\{ myProperty: string; myPropertyTwo?: number \\\\| undefined; \\\\} | \\\\{ myProperty: string; \\\\} |",
      "",
      "### Return type",
      "",
      "T",
      "",
      "",
      "",
    ],
  },
  Object {
    "FileLocation": "index\\\\foo.md",
    "Result": Array [
      "## class: Foo",
      "",
    ],
  },
  Object {
    "FileLocation": "index\\\\world.md",
    "Result": Array [
      "## class: World",
      "",
    ],
  },
  Object {
    "FileLocation": "index\\\\earth.md",
    "Result": Array [
      "## class: Earth",
      "",
    ],
  },
]
`;
